packages/firestore/test/integration/remote/stream.test.ts

```diff
@@ -0,0 +1,217 @@
+   */
+  verifyNoPendingCallbacks(): void {
+    expect(this.pendingCallbacks).to.be.empty;
+  }
+
+  onHandshakeComplete(): Promise<void> {
+    return this.resolvePending('handshakeComplete');
+  }
+
+  onMutationResult(
+    commitVersion: SnapshotVersion,
+    results: MutationResult[]
+  ): Promise<void> {
+    return this.resolvePending('mutationResult');
+  }
+
+  onWatchChange(
+    watchChange:
+      | DocumentWatchChange
+      | WatchTargetChange
+      | ExistenceFilterChange,
+    snapshot: SnapshotVersion
+  ): Promise<void> {
+    return this.resolvePending('watchChange');
+  }
+
+  onOpen(): Promise<void> {
+    return this.resolvePending('open');
+  }
+
+  onClose(err?: firestore.FirestoreError): Promise<void> {
+    return this.resolvePending('close');
+  }
+
+  private resolvePending(actualCallback: StreamEventType): Promise<void> {
+    if (this.pendingPromises.length > 0) {
+      let pendingPromise = this.pendingPromises.shift();
+      pendingPromise.resolve(actualCallback);
+    } else {
+      this.pendingCallbacks.push(actualCallback);
+    }
+    return Promise.resolve();
+  }
+}
+
+describe('Watch Stream', () => {
+  let streamListener: StreamStatusListener;
+
+  beforeEach(() => {
+    streamListener = new StreamStatusListener();
+  });
+
+  afterEach(() => {
+    streamListener.verifyNoPendingCallbacks();
+  });
+
+  /**
+   * Verifies that the watch stream does not issue an onClose callback after a
+   * call to stop().
+   */
+  asyncIt('can be stopped before handshake', () => {
+    let watchStream: PersistentListenStream;
+
+    return withTestDatastore(ds => {
+      watchStream = ds.newPersistentWatchStream(streamListener);
+      watchStream.start();
+
+      return streamListener.awaitCallback('open').then(() => {
+        // Stop must not call onClose because the full implementation of the delegate could
+        // attempt to restart the stream in the event it had pending watches.
+        watchStream.stop();
+      });
+    });
+  });
+});
+
+describe('Write Stream', () => {
+  let queue: AsyncQueue;
+  let streamListener: StreamStatusListener;
+
+  beforeEach(() => {
+    queue = new AsyncQueue();
+    streamListener = new StreamStatusListener();
+  });
+
+  afterEach(() => {
+    streamListener.verifyNoPendingCallbacks();
+  });
+
+  /**
+   * Verifies that the write stream does not issue an onClose callback after a
+   * call to stop().
+   */
+  asyncIt('can be stopped before handshake', () => {
+    let writeStream: PersistentWriteStream;
+
+    return withTestDatastore(ds => {
+      writeStream = ds.newPersistentWriteStream(streamListener);
+      writeStream.start();
+      return streamListener.awaitCallback('open');
+    }).then(() => {
+      // Don't start the handshake.
+
+      // Stop must not call onClose because the full implementation of the delegate could
+      // attempt to restart the stream in the event it had pending watches.
```

Comment:
delagate => callback and while you're here maybe fix "pending watches" to "pending writes" (even though Obj-C is wrong too :-))

Reply:
Done.